### PR TITLE
update timing testing documentation with lessons learned

### PR DESCRIPTION
**Pull request by tomato42**
### Motivation and Context

make the documentation consistent with current code and the existence of `test-bleichenbacher-timing-pregenerate.py` script

### Checklist

- [x] I have read the [CONTRIBUTING.md](https://github.com/tomato42/tlsfuzzer/blob/master/CONTRIBUTING.md) document and my PR follows [change requirements](https://github.com/tomato42/tlsfuzzer/blob/master/CONTRIBUTING.md#change-requirements) therein
- [x] the changes are also reflected in documentation and code comments
- [x] all new and existing tests pass (see CI results)
- [ ] [test script checklist](https://github.com/tomato42/tlsfuzzer/wiki/Test-script-checklist) was followed for new scripts
- [ ] new test script added to `tlslite-ng.json` and `tlslite-ng-random-subset.json`
- [x] new and modified scripts were ran against popular TLS implementations:
  - n/a
- [x] required version of tlslite-ng updated in requirements.txt and README.md

- - -
This change is [Reviewable](https://reviewable.io/reviews/tlsfuzzer/tlsfuzzer/858)

### CODE REVIEW

File: docs/source/timing-analysis.rst [hunk 1]
```diff
@@ -219,32 +259,30 @@ samples in question), but it's a good starting point.
 Also, it means that if you wish to decrease the reported confidence interval
 by a factor of 10, you must execute the script with 100 times as many
 repetitions (as 10²=100).
+Or execute the same script with the same settings 100 times, combine
+the resulting data in ``timing.csv`` files and analyse such combined data
+set.
+That's the primary reason for the careful system setup: it's much
+easier to adjust a system configuration that to execute hundred tests
```
Comment (ZoltanFridrich) [inline docs/source/timing-analysis.rst:266]: `than to execute` instead of `that to execute`

File: docs/source/timing-analysis.rst [hunk 2]
```diff
@@ -258,7 +296,7 @@ To run the tests:
 1. Select a machine with sufficient cooling and a multi-core CPU
 2. Use methods mentioned before to create isolated cores, watch out for
    hyperthreading
-3. For RSA tests use small key (1024 bit), for CBC tests use a fast cipher and
+3. For RSA tests use small key (1024 bit), use a fast cipher and
```
Comment (ZoltanFridrich) [inline docs/source/timing-analysis.rst:299]: Either remove `use a` or start a new sentence. This way it sounds weird.

File: docs/source/timing-analysis.rst [hunk 3]
```diff
@@ -436,19 +477,27 @@ analysis as usual.
...
 the data is
 `autocorrelated
 <https://en.wikipedia.org/wiki/Autocorrelation>`_ (or, in other words,
 not-independent) and simple summary statistics like
 mean, median, or quartiles are not representative of the samples.
+It also means that they can't be compared with the box test, or Mann-Whitney
+U test.
 
-The next set of graphs show the overall shape of the samples.
+The next set of graphs compares the overall shape of the samples.
```
Comment (ZoltanFridrich) [inline docs/source/timing-analysis.rst:500]: `compare` instead of `compares`

File: docs/source/timing-analysis.rst [hunk 4]
```diff
@@ -474,36 +523,40 @@ The ``diff_ecdf_plot.png`` is the ECDF counterpart to the scatter plot.
...
 <https://en.wikipedia.org/wiki/Bootstrapping_(statistics)>`_ confidence
 interval for them.
 For an implementation without a timing side channel present, all the graphs
 should intersect with the horizonal 0 line.
-If a graph does not intersect with the 0 line, then the number of heights
-of it from the 0 line suggests how strong is the confidence in the
-presence of side channel on an exponential scale.
+If a graph does not intersect with the 0 line, then the distance of it
+from the 0 line suggests how strong is the confidence in the
+presence of side channel is on an exponential scale.
```
Comment (ZoltanFridrich) [inline docs/source/timing-analysis.rst:547]: This whole sentence is difficult to read. Can this be split or rewritten so it is more understandable? Or maybe some commas would help.

**Review by ZoltanFridrich — CHANGES_REQUESTED** [review]
The text requires some fixes.

File: docs/source/timing-analysis.rst [hunk 5]
```diff
@@ -146,67 +163,90 @@ And the general requirements to collect and analyse timing results:
...
-That means that the scripts need to use statistical tests to check if the
+and almost never `normal <https://en.wikipedia.org/wiki/Normal_distribution>`_.
+Moreover, the measurements are almost always self-similar (because when
+a CPU starts to run on a different frequency *all* measurements will be
+affected by it),
+thus statistical tests used must not require
+`independent and identically distributed
+<https://en.wikipedia.org/wiki/Independent_and_identically_distributed_random_variables>`_
+measurements.
+That means that the scripts need to use statistical tests suited to testing
```
[the remark this answers is not in the document]
Comment (tomato42) [reply docs/source/timing-analysis.rst:198]: done

File: docs/source/timing-analysis.rst [hunk 6]
```diff
@@ -146,67 +163,90 @@ And the general requirements to collect and analyse timing results:
...
 After executing it against two sets of observations (samples), it outputs
 a "p-value"—a probability of getting such samples, if they were taken from
 the same population.
-A high p-value (close to 1) means that the samples likely came from the
-same source while a small value (close to 0, smaller than 0.05) means
-that it's unlikely that they came from the same source distribution.
+A high p-value (close to 1, larger than 0.05) means that the samples likely
+came from the
+same source while the smalller the value (closer to 0, smaller than 0.05) the
+more likely it is that they don't come from the same same source distribution.
```
[the remark this answers is not in the document]
Comment (tomato42) [reply docs/source/timing-analysis.rst:215]: done

File: docs/source/timing-analysis.rst [hunk 7]
```diff
@@ -146,67 +163,90 @@ And the general requirements to collect and analyse timing results:
...
+<https://en.wikipedia.org/wiki/Sign_test>`_ to compare pairs of samples.
 After executing it against two sets of observations (samples), it outputs
 a "p-value"—a probability of getting such samples, if they were taken from
 the same population.
-A high p-value (close to 1) means that the samples likely came from the
-same source while a small value (close to 0, smaller than 0.05) means
-that it's unlikely that they came from the same source distribution.
+A high p-value (close to 1, larger than 0.05) means that the samples likely
+came from the
+same source while the smalller the value (closer to 0, smaller than 0.05) the
```
[the remark this answers is not in the document]
Comment (tomato42) [reply docs/source/timing-analysis.rst:214]: done

File: docs/source/timing-analysis.rst [hunk 8]
```diff
@@ -10,18 +10,24 @@ When an implementation takes different amounts of time to process the messages
...
+<https://github.com/tomato42/tlsfuzzer/blob/master/scripts/test-bleichenbacher-timing-pregenerate.py>`_
 script.
 
-The other is related to de-padding and verifying MAC values in CBC ciphertexts,
-the newest iteration of which is called Lucky Thirteen. You can test for it
-using the
-`test-lucky13.py
-<https://github.com/tomato42/tlsfuzzer/blob/master/scripts/test-lucky13.py>`_
-script.
+While we also include another script to do Bleichenbacher side-channel
```
[the remark this answers is not in the document]
Comment (tomato42) [reply docs/source/timing-analysis.rst:20]: no, `while` is necessary for the `similarly robust approach` to make sense

File: docs/source/timing-analysis.rst [hunk 9]
```diff
@@ -10,18 +10,24 @@ When an implementation takes different amounts of time to process the messages
...
+testing
+(`test-bleichenbacher-timing.py
+<https://github.com/tomato42/tlsfuzzer/blob/master/scripts/test-bleichenbacher-timing.py>`_)
+and one to test de-padding and verifying MAC values in CBC ciphertexts
+(`test-lucky13.py
+<https://github.com/tomato42/tlsfuzzer/blob/master/scripts/test-lucky13.py>`_,
+the Lucky Thirteen attack), they do not use similarly robust approach
```
[the remark this answers is not in the document]
Comment (tomato42) [reply docs/source/timing-analysis.rst:27]: no, same as the `While` in the same sentence

File: docs/source/timing-analysis.rst [hunk 10]
```diff
@@ -146,67 +163,90 @@ And the general requirements to collect and analyse timing results:
...
 <https://en.wikipedia.org/wiki/Uniform_distribution_(continuous)>`_ with
-values between 0 and 1.
-
-You can use this property to check if not only the failures (small p-values)
-occur not more often than expected, but to check for more general inconsistency
-in p-values (as higher probability of small p-values means that large
-p-values occur less often).
-
-The scripts perform the
+values between 0 and 1. It's thus then possible, by using external statistical
```
[the remark this answers is not in the document]
Comment (tomato42) [reply docs/source/timing-analysis.rst:233]: done

File: docs/source/timing-analysis.rst (hunk 2 above)
Comment (tomato42) [reply docs/source/timing-analysis.rst:299]: done

File: docs/source/timing-analysis.rst (hunk 4 above)
Comment (tomato42) [reply docs/source/timing-analysis.rst:547]: rewritten, please check

File: docs/source/timing-analysis.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (ZoltanFridrich) [reply docs/source/timing-analysis.rst:20]: ok

File: docs/source/timing-analysis.rst (hunk 9 above)
[the remark this answers is not in the document]
Comment (ZoltanFridrich) [reply docs/source/timing-analysis.rst:27]: alright

File: docs/source/timing-analysis.rst [hunk 11]
```diff
@@ -31,6 +37,10 @@ a server running alone on a machine, with no interruptions from other
 services or processes will provide statistically significant results with
 fewest observations.
 
+It should be noted though, that using special setup for the system on which
+you run the test will affect *how long* the test needs to run, it will *not*
```
[the remark this answers is not in the document]
Comment (ZoltanFridrich) [reply docs/source/timing-analysis.rst:41]: Yeah, I guess I meant running time or something like that. Its fine.

File: docs/source/timing-analysis.rst (hunk 4 above)
Comment (ZoltanFridrich) [reply docs/source/timing-analysis.rst]: looks good

**Review by ZoltanFridrich — APPROVED** [review]
LGTM!

**Pull request merged**